File: 3. Arrays/jump_game.py

class Solution:

    def canJump(self, nums: List[int]) -> bool:
        leftMost = len(nums) - 1
        for x in range(len(nums) - 1, -1, -1):
            # for x in range(len(nums)):
            # fromBack = -x - 1
            if (x + nums[x] >= leftMost):
                leftMost = x

        return leftMost == 0


class Solution:

    def canJump(self, nums: List[int]) -> bool:
        rightMost = 0
        for idx in range(len(nums)):
            currJump = idx + nums[idx]
            if (idx > rightMost):
                return False
            if (currJump > rightMost):
                rightMost = currJump
        return True

File: 3. Arrays/jumpGame.js
// Greedy
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
    let leftMost = nums.length - 1;
    for (let i = nums.length - 1; i >= 0; i--) {
        if (i + nums[i] >= leftMost) {
            leftMost = i;
        }
    }
    return leftMost === 0;
};


// Greedy
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
    let rightMost = 0;
    for (let idx = 0; idx < nums.length; idx++){
      let currJump = idx + nums[idx];
      if (idx > rightMost) return false;
      if (currJump > rightMost)
        rightMost =  currJump
    }
    return true
};
